src/openai/_models.py

```diff
@@ -356,7 +357,10 @@ def _construct_field(value: object, field: FieldInfo, key: str) -> object:
         return field_get_default(field)
 
     if PYDANTIC_V2:
-        type_ = field.annotation
+        if field.metadata:
+            type_ = wrap_in_annotated_type(field)
+        else:
+            type_ = field.annotation
```

Comment:
In the case where the response data for a union field doesn't conform to expected (i.e. we fail the validate_type call), we seem to lose the metadata (which has the discriminator info) when constructing the field. This makes the code construct the first type present in the union instead of creating the actual known type. Test case illustrating this [here](https://github.com/openai/openai-python/blob/126912e307622106b537ed4d578cbf2af16180e0/tests/test_models.py#L832)